pagefaultgames/pokerogue
# [Refactor] Add type inference and support for simulated calls to `ArenaTag.apply`

## What are the changes the user will see?
N/A

## Why am I making these changes?
Simulated calls to arena tags are necessary for the Enemy AI to evaluate turn actions silently, especially when [Infiltrator](#4636) is implemented. With that change, I figured it was also a good time to add type inference to `ArenaTag.apply` along the lines of #4294 

## What are the changes from a developer perspective?
`data/arena-tag`: The signature for `ArenaTag.apply` has changed from
```ts
apply(arena: Arena, args: any[]);
```
to
```ts
apply(arena: Arena, simulated: boolean, ...args: unknown[]);
```

The new `simulated` argument, much like the argument of the same name for abilities' `apply` functions, allows arena tags to be applied without any observable changes to game state. `args` has also been changed to use spread syntax, which allows `overrides` of the function in subclasses to expand `args` into multiple arguments.

`applyTagsOnSide` and `applyTags` have also been changed accordingly, along with all references to the two functions in the codebase.

### Screenshots/Videos
N/A

## How to test the changes?
For a specific example, you can check the screen tests:
`npm run test reflect`
`npm run test light_screen`
`npm run test aurora_veil`

## Checklist
- [n/a] Are the changes visual?
  - [n/a] Have I provided screenshots/videos of the changes?
